Redesign the main pattern inserter

youknowriad opened the pull request:
## What?

See #41379

This PR redesigns the patterns inserter. The first step is just to show the categories before actually rendering the patterns.
Here's the current state

https://user-images.githubusercontent.com/272444/189339119-c3f787ad-65ae-4af9-af35-3e1936c15c3b.mov

As you can see it's still very rough but we can get an idea of what we're trying to achieve, here are some of the remaining challenges:

- [x] Address keyboard navigation properly.
- [x] Potentially render a different component in mobile (Navigation pattern like the styles sidebar).
- [x] Design tweaks: It's proving challenging to position the "explore all" button at the bottom of the inserter consistently.
- [x] Tests.

## Testing Instructions

 - Open the inserter
 - Open the patterns tab
 - Play with it

youknowriad committed:
Redesign the main pattern inserter

youknowriad committed:
Use navigation pattern on mobile

youknowriad committed:
Keyboard navigation improvements

youknowriad committed:
Add a shadow and a subtle animation to the patterns panel

jasmussen commented:
This is glorious, and I look forward to reviewing and helping in any way I can. This would be excellent to land soon!

I'm encountering this issue testing the branch:

```
Fatal error: Cannot redeclare wp_add_global_styles_for_blocks() (previously declared in /Users/joen/Local Sites/local-wordpress/app/public/wp-includes/global-styles-and-settings.php:202) in /Users/joen/GIT/gutenberg/lib/compat/wordpress-6.1/get-global-styles-and-settings.php on line 11
```

<img width="1392" alt="Screenshot 2022-09-12 at 11 26 46" src="https://user-images.githubusercontent.com/1204802/189619709-2d73043b-0519-4220-b7f5-ba463f73aaa1.png">

Any thoughts?

youknowriad commented:
@jasmussen I think that issue is unrelated, probably due to some on going core merges. So try to destroy your environment or use WP 6.0 as a base WP install.

jasmussen commented:
As an environmentalist, I prefer not to destroy the environment 😜 — however I was able to downgrade, and yes, 6.0 works great. Diving in, thanks.

jasmussen commented:
As a single PR, this feels incredible already:

![status](https://user-images.githubusercontent.com/1204802/189621464-b11898a5-d971-4da2-ad43-0eac0042a960.gif)

A few small tweaks I think we can make. The "Explore all patterns" button:

<img width="349" alt="Screenshot 2022-09-12 at 11 35 08" src="https://user-images.githubusercontent.com/1204802/189621619-54619ea4-454a-4393-9365-35757d45b6c7.png">

can we make it more like the "Manage your library" button here? 

<img width="624" alt="Screenshot 2022-09-12 at 11 35 29" src="https://user-images.githubusercontent.com/1204802/189621654-a1a4031a-d34a-4844-9d04-92ba58cf1fdc.png">

The button style feels like a nice bookend to the list.

I happen to be running a light-gray theme, so the light gray background makes things blend in here a bit:

<img width="994" alt="Screenshot 2022-09-12 at 11 36 31" src="https://user-images.githubusercontent.com/1204802/189621813-93ce5c2e-ee7b-413f-9a1d-3959514546e0.png">

I'm going to explore what we can do here to separate things out a bit better.

One open question is whether we can improve some of the space usage. I think it's okay that the pattern flyout covers content, but I wonder if we could reduce the width of the inserter when the pattern is opened 🤔

jasmussen commented:
Alright a couple of quick mockups.

<img width="3000" alt="V1" src="https://user-images.githubusercontent.com/1204802/189626941-7aa99a13-341a-449e-9158-a7260c30eaed.png">

* Shows how the background color of the overlay actually isn't a gray color, but a 95% opaque white color with a `backdrop-filter: blur(8px)`.
* Adds a gray border left of the flyout.
* Has 2px border radii around each pattern preview
* Nice to have: shows how the default 350px inserter scales down whe the flyout is opened.

One cool thing would be to leverage the recently merged zooming, and invoke that when this pattern library is opened:

<img width="3000" alt="V2" src="https://user-images.githubusercontent.com/1204802/189627255-eaaa21d4-3741-41b2-8e09-163a8db39398.png">

Let me know if you'd like me to try and push some of the border/background-color/filter tweaks!

youknowriad commented:
@jasmussen Feel free to push the design tweaks.

> can we make it more like the "Manage your library" button here?

I tried this but the way the inserter is styled / organized, it's not straight forward. I'll get it another try though.

jasmussen committed:
Polish

jasmussen commented:
I pushed a little bit of polish around paddings, border, background colors, styles. It now looks like this:

![update](https://user-images.githubusercontent.com/1204802/189639319-a34d2d82-c205-4aa1-afcf-5bcf7ef9134c.gif)

This adds a backdrop-filter which works decently well for making the patterns feel separate from the content, yet still overlay them. However I want to note a performance concern here, blur is expensive to animate. Seems like we could either remove the blur (the blur especially becomes useless if we zoom out), or we can remove the initial animation, which would also be unfortunate. Any alternatives?

One thought on the animation, though — right now it's a bit bouncy. If we can choose a "ease-out" that would likely feel better.

Did you have any thoughts on whether we could animate the main inserter to be smaller?

youknowriad commented:
>  blur is expensive to animate. Seems like we could either remove the blur (the blur especially becomes useless if we zoom out)

What's the "blur" that you're talking about here?

> One thought on the animation, though — right now it's a bit bouncy. If we can choose a "ease-out" that would likely feel better.

Yes, we can tweak it, that's just the default slide in from framer motion.

> Did you have any thoughts on whether we could animate the main inserter to be smaller?

I'm hopeful that we can do that. I'll try a bit later. I'm focused on this for now #44088 which is a blocker to move the "browse patterns" to the bottom.

jasmussen commented:
> What's the "blur" that you're talking about here?

I lowered the opacity of the white bg color to make this more visible:

<img width="1392" alt="Screenshot 2022-09-12 at 16 04 14" src="https://user-images.githubusercontent.com/1204802/189674541-6adfb19a-9013-4bf0-8ec0-572dfe32d2f7.png">

It's mostly visible when the pattern flyout covers a wide range of content. We want to find a balance between the blur being visible and the content inside being legible. We can also drop that blur. CC: @SaxonF.

youknowriad commented:
I tested the last changes a bit, I personally preferred the gray background over the blurry white, especially in white bg themes  (which I think are probably the most common). I also like the contrast between the inserter tabs (white) and the patterns list (gray)

jasmussen committed:
Restore gray bg.

jasmussen commented:
Alright, I restored the gray background for now, we can always revisit:

<img width="1392" alt="Screenshot 2022-09-13 at 08 36 01" src="https://user-images.githubusercontent.com/1204802/189827813-bc1f6d12-50b6-4fbc-988e-777e4c547c1c.png">

jameskoster commented:
Quick note that this isn't working properly in the Site Editor. The pattern panel doesn't appear when clicking a category, unless you first navigate _away_ from the Patterns tab (to Blocks for example) in the Inserter.

jameskoster commented:
A right border could help separate the pattern panel from the canvas?

<img width="823" alt="Screenshot 2022-09-13 at 10 45 09" src="https://user-images.githubusercontent.com/846565/189869547-a762d49c-e3c2-4f69-bc99-d7ed1592ad3a.png">

jasmussen committed:
Add right border.

jasmussen commented:
Agreed. Let's add the right border for now:
<img width="697" alt="Screenshot 2022-09-13 at 11 48 55" src="https://user-images.githubusercontent.com/1204802/189870438-0bbb734e-00fc-4abf-b27f-6978ad1a5d5e.png">

We can always revisit.

youknowriad commented:
> Quick note that this isn't working properly in the Site Editor. The pattern panel doesn't appear when clicking a category, unless you first navigate away from the Patterns tab (to Blocks for example) in the Inserter.

I was not able to reproduce this.

jameskoster commented:
Here's what I'm seeing in case it helps:


https://user-images.githubusercontent.com/846565/189886970-edbbd639-0feb-4c8c-aa38-765c9a6e9706.mp4

youknowriad commented:
@jameskoster yeah nvm, I managed to break it now :)

youknowriad committed:
Add width animation

youknowriad committed:
Fix site editor patterns category selection

youknowriad committed:
Fix inserter width on mobile

youknowriad commented:
@jameskoster the bug should be fixed.

@jasmussen The inserter scales down a bit now

jasmussen commented:
This is excellent!

![status](https://user-images.githubusercontent.com/1204802/189894662-b42f9fc2-d6f3-4d72-84f6-2e1b8ce4aa57.gif)

I notice that there's also animation on the sidebar as it appears initially. The animation from the left side seems okay enough to indicate directionality, and it's a nice little bit of flair. However it would be good to test the performance of this with very long and complex pages, as it essentially also animates the viewport/iframe itself — I don't know if that should be expensive, it might not be? But worth testing.

But in any case, we should replace the bouncy animation with a simpler ease-out, or it becomes a bit disorienting. 

Really nice work, I honestly think we can land this in short order! 🔥

youknowriad commented:
The only thing that I'm uncertain about and to be honest, not sure how to solve is keyboard navigation. What should be the behavior here?

jasmussen commented:
Would arrow key up/down and left/right work?

https://user-images.githubusercontent.com/1204802/189896206-dcbbd1a1-330f-472a-96f0-e9af76bd5066.mov

youknowriad commented:
@jasmussen I think that's probably the best option but the code will be pretty ugly because the design is not straightforward like regular menus, the panel and the list are a bit separated in the components tree...

jasmussen commented:
Would a modal behavior be simpler? I.e. click the category and focus is set in the flyout?

youknowriad commented:
> Would a modal behavior be simpler? I.e. click the category and focus is set in the flyout?

Simpler to implement yes, that's what I've already tried but there's something broken right now. I think

youknowriad committed:
Move the explore patterns button to the bottom

youknowriad commented:
I've moved the "explore patterns" button to the bottom of the panel

youknowriad committed:
Reduce the transition duration

youknowriad commented:
I tried a different transition duration, let me know if the animation is better now.

jasmussen commented:
Button looks good. I'd love to polish it to be the same style as the "Style variation picker" button, but I for simplicity we can keep it like this for now.

The animation looks much better, nice!

Modal behavior for the flyout seems good, though indeed there does seem to be something buggy with it, pressing escape doesn't close it, and it seems like I have to press tab only once and then use arrow keys, lest the thing closes. This seems like the final thing we need to fix then this can land.

youknowriad commented:
> I'd love to polish it to be the same style as the "Style variation picker" button

Could we maybe formalize this into a variant or something to avoid reinventing new styles for each UI piece.

jasmussen commented:
Exactly why I think we need to look at that separately 👍 👍

javierarce commented:
Fantastic work on this feature! 

There's something I think we should address at some point which is how we load the previous: it would be great if we could have some smooth loading transitions instead of the current shaky behavior. 

Here's an [old issue](https://github.com/WordPress/gutenberg/issues/36491) that I think it's still valid.

youknowriad committed:
Enhance keyboard navigation

youknowriad committed:
Better nested constrained tabbing support

youknowriad commented:
I improved the keyboard navigation, I think it's more decent now but it would benefit from more screen reader testing.
I don't love the current code I wrote and behavior but it's something that we can ship IMO.

youknowriad commented:
> There's something I think we should address at some point which is how we load the previous: it would be great if we could have some smooth loading transitions instead of the current shaky behavior.

@javierarce Oh yes, this is indeed not great in some situations. It's a very hard problem to solve though. I'm interested in doing more exploration and research though but without promises :).

alexstine commented:
Adding to my reviews.

mtias committed:
Add shadow to pattern items and remove title from structure.

youknowriad commented on packages/block-editor/src/components/block-patterns-list/index.js:
```diff
@@ -45,9 +45,6 @@ function BlockPattern( { isDraggable, pattern, onClick, composite } ) {
 						/>
-						<div className="block-editor-block-patterns-list__item-title">
-							{ pattern.title }
-						</div>
```
@mtias Some patterns have descriptions that was already rendered right after the title? Should it be removed as well?

Also, I guess this removal means we need to add a label to the `CompositeItem` component for screen readers.

youknowriad replied:
Oh I just saw that there's already a label in its wrapper div, so maybe it's good enough already. We need to test a screen reader.

mtias replied:
It's a pending design task to explore if a tooltip is helpful too

jameskoster commented on packages/block-editor/src/components/block-patterns-list/style.scss:
```diff
@@ -18,8 +18,9 @@
-		border-radius: $radius-block-ui + $border-width;
+		border-radius: 4px;
 		border: $border-width solid $gray-100;
+		box-shadow: 0 15px 25px rgb(0 0 0 / 7%);
```
Should we make this a var [like we did with modal/popover shadows](https://github.com/WordPress/gutenberg/blob/3a2a5559b60540ae2098b633f9b54d7676a61ba5/packages/base-styles/_variables.scss#L66-L67)? There may be other places in the UI where we want to use the same effect.

jasmussen replied:
I'd love some system for shadows, a pet peeve of mine is to "obey the laws of physics". This is a very soft shadow, suggesting anything above it should be even softer, i.e. modals.

jameskoster replied:
There is an [experimental elevation](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-elevation--default) component, but I don't know if it's used anywhere.

mtias committed:
Make drag chip read "Pattern" when dragging a pattern.

jameskoster commented:
Can we do something with empty categories?

<img width="600" alt="Screenshot 2022-09-14 at 18 24 43" src="https://user-images.githubusercontent.com/846565/190221686-ae725c9d-32a2-48ae-9533-bdf5e9682a53.png">

A "No patterns found", or maybe just hide the category altogether?

youknowriad commented:
@jameskoster it should already be the case, maybe you have a pattern there but its height is `0` ?

youknowriad commented:
Tested a bit with VoiceOver, it seems decent though my experience with these tools is small. 

What's remaining here?

jameskoster commented:
> @jameskoster it should already be the case, maybe you have a pattern there but its height is 0 ?

Okay so there _are_ patterns in the category. They appear on trunk, and on this PR in the pattern explorer:

<img width="464" alt="Screenshot 2022-09-15 at 11 07 17" src="https://user-images.githubusercontent.com/846565/190377487-fc7226e4-a077-4005-ae86-ea7b7ea03b08.png">
<img width="1230" alt="Screenshot 2022-09-15 at 11 08 06" src="https://user-images.githubusercontent.com/846565/190377490-f04980d1-7546-41dc-b82c-043a1bf16d48.png">

(Should the new pattern preview with shadow be leaking into the explorer?)

But for some reason the inserter panel is entirely empty:

<img width="1280" alt="Screenshot 2022-09-15 at 11 10 14" src="https://user-images.githubusercontent.com/846565/190377671-9980dd62-154b-4778-a829-14f8f6cd1d83.png">

All the other categories are working fine. 

---

Taking the screenshot above I noticed we need to do something about the mobile experience too :)

youknowriad committed:
Fix uncategorized patterns

youknowriad commented:
@jameskoster Yeah, there was a bug for patterns with a list of categories that are unregistered, it should be fixed now.

youknowriad committed:
Add scrollbar gutter

youknowriad commented:
@jameskoster  @jasmussen to avoid the jumps when the patterns are being loaded, I've added a `scrollbar-gutter` CSS that is useful when the scrollbars are always visible (windows and an option in Mac OS too) and I adjusted the padding a bit to try to optimize it for both situations (when the option is enabled or disabled), let me know how it feels.

Note: scrollbar-gutter is not supported yet in Safari (Chrome and Firefox only)

jameskoster commented:
**Before**

https://user-images.githubusercontent.com/846565/190389295-9accf0dd-6819-4a78-948b-77d354985d2f.mp4

**After**

https://user-images.githubusercontent.com/846565/190389289-1379037e-af04-434e-8422-d188160a8e30.mp4

Seems like a solid improvement to me.

mtias committed:
Try a stronger border outline for selecting patterns.

youknowriad commented:
Just noting that the design tweaks made here to pattern lists: shadow and title-less patterns are currently applied to all places where patterns are shown:

 - Quick inserter (site editor top level)
 - Template part creation

Not sure if this was intended?

youknowriad committed:
Apply the new pattern preview style to the main inserter only

youknowriad commented:
In 8d1fa31 I've limited the title-less + shadow design to the main inserter only leaving all the rest of the UI untouched.

mtias committed:
Simplify border shape and use dark gray.

mtias committed:
Use Heading component for category label.

mtias committed:
Restrict hover effects to main inserter.

youknowriad committed:
Avoid animations and width changes for now

youknowriad commented:
I've removed the animations and width changes for now (a bit fragile specially in Safari). We can consider bringing some of this back later if we find a solid implementation.

jasmussen commented:
Can we keep the width changes, just without any animation? I would imagine it's especially the canvas animating that's tricky.

youknowriad commented:
> Can we keep the width changes, just without any animation?

We can, any thoughts @mtias

mcsf commented on packages/compose/src/hooks/use-constrained-tabbing/index.js:
```diff
@@ -38,10 +38,16 @@ function useConstrainedTabbing() {
 
+			// constrainedTabAlreadyHandled is a custom property added
+			// to event to support nested constrainted tabbing containers
+			// @ts-ignore
```
Comment no longer relevant.

mcsf commented on packages/block-editor/src/components/block-patterns-list/index.js:
```diff
@@ -22,7 +22,11 @@ function BlockPattern( { isDraggable, pattern, onClick, composite } ) {
+		<InserterDraggableBlocks
+			isEnabled={ isDraggable }
+			blocks={ blocks }
+			isPattern={ !! pattern }
```
When would `!! pattern` not be true?

mcsf commented on packages/block-editor/src/components/inserter/menu.js:
```diff
@@ -241,6 +247,23 @@ function InserterMenu(
+					category={ selectedPatternCategory }
+					onClose={ () => {
+						if (
+							searchRef.current.ownerDocument.activeElement.classList.contains(
+								'block-editor-inserter__patterns-category'
+							)
```
Not really the point of this PR, but the setup with `searchRef`, `focusSearch` and the other places where that string occurs, is hard to decipher.

mcsf commented on packages/block-editor/src/components/inserter/block-patterns-tab.js:
```diff
@@ -138,35 +84,204 @@ function BlockPatternsTabs( {
+	} );
+
+	useEffect( () => {
+		container?.current?.focus();
```
Is the `container` ref needed? What's the difference between this and `ref?.current?.focus()`?

youknowriad replied:
Ref is returned from useDialog, we don't control its shape here so it could be an object with "current" or it could be a callback (functions are valid refs). So Adding our own self controlled ref is safer.

mcsf reviewed (commented):
Looking great. :)

jameskoster commented:
Should we remove the 'Explore all patterns' button for now? Until it's connected with the Pattern Directory it's not all that useful.

jasmussen commented:
I think we should keep it as is, lest people will ask "where did that go?"

alexstine commented:
@youknowriad First off, sorry for the horrible delay on my end. I wish there were more people from Accessibility Team testing these PRs.

Anyway, this is super inaccessible. 😞 Here is a list to get started.
1. After selecting the Patterns tab, I am presented with a long list of buttons. Is there any way to render those as filtering tabs instead of buttons? Just like Blocks and Patterns?
2. Once you do make a selection, something weird happens. The page nearly crashes, so I am guessing there is still some outstanding work to be done on performance. However, focus seems to completely get lost.
```
document.activeElement;
<div class="block-editor-inserter__patterns-category-panel" tabindex="-1">
```
3. Once you tab in to the list of patterns, there is no way to shift+tab out again.
4. When using arrow keys to explore the patterns list, all I hear is "list item, list item, list item". Not really useful, this needs major context adjustment.

This does not even come close to being core ready so I appreciate you waiting on me to check. Please let me know if I can be of further assistance or when you need another review.

Thanks.

youknowriad committed:
Add label for the category patterns tab

youknowriad commented:
Thanks for testing @alexstine that's very helpful.

So my idea right now was to consider the "patterns" tab as a menu where you select a category and once you click a category it opens a "dialog" where you can navigate the patterns of this category. (So in the dialog, tabbing is trapped, you hit escape to go back...). I think I was missing a label on the dialog container (that's where you thought that the focus was lost but it wasn't, it was not labeled properly though).

So maybe I can add a label to the dialog (already did), and make the category list a "menu". WDYT?

---
The alternative you suggest is also interesting: make the category list a "tablist" and the panel a "tab panel". I can try that but the difficulty here is that the "tab panel" can't be rendered right after the "tab list" in the DOM, so I still need to move the focus manually when you select a tab... and have it work like a dialog maybe?

alexstine commented:
@youknowriad Still not working and I pulled down latest changes.

It looks like what you are calling a dialog is actually some type of ARIA list with an iFrame. If I select Text button for example, I see a heading level 4 that has Text and right below that is a list.

If you were planning on using a real dialog, you need to use something like the `useDialog` hook from wordpress/compose.

How are you currently implementing the patterns to choose from?

What if you used a tab list but disabled the switching on tab navigate? E.g. The only way to select a tab is via arrow keys then space/enter to activate. That way the tab would only load when necessary. Thoughts?

Thanks.

alexstine commented:
@youknowriad Okay, here is a recording of what it is like with a screen reader.

https://drive.google.com/file/d/1HC3vp6AiIGTcr8az402bZO1Dab5GbUEm/view?usp=sharing

Sorry, it is nearly midnight here when I recorded it so I kind of rambled on. Hopefully something in that is useful other than me telling you it just does not work. Hopefully, based off what you can see with my interaction, you can either fix it or explain to me the goal and maybe I can help more.

Thanks.

youknowriad commented:
@alexstine Thanks for the great feedback and video, that's very helpful. I've been AFK for a few days so I was unable to reply here sooner.

> If you were planning on using a real dialog, you need to use something like the useDialog hook from wordpress/compose.

I'm already using useDialog for the dialog, but I noticed in your video (it works for me) that focus is not trapped in the dialog. The dialog actually contains two things: a heading and a list of patterns to pick from.

Inside the list item, I think we're supposed to use arrow keys to navigate between the list items, I noticed on your video that the list items are not announced properly but the thing is, we didn't really change anything in this PR for this list aside hiding the "title" visually, the label is still applied, but yeas it seems it's applied to the wrong element (the parent element of the list item instead of the list item directly, I'll see if moving this fixes it)

The other thing that you mentionned: tabs not working is I think due to the fact that the dialog only contains one tabbable item (the list of patterns) and since useDialog adds constrained tabbing inside the dialog, tabbing just loops, so it stays on the same and unique tabbable element of the dialog.

youknowriad committed:
Move the aria label for the list items to the correct element

youknowriad commented:
I've moved the label to the "list item" element, let me know if it improves things. I noticed that in Voiceover, whether the aria-label is on the list item or its parent div doesn't impact what gets announced but it's probably different for you on windows.

alexstine commented:
@youknowriad Okay, so we're getting closer. That `aria-label` fix did at least make things make sense. However, I think using a dialog in this way may not be correct. Can you test and see how the interaction is with the list minus the dialog? The problem is, once you select a category, that is it. There is no way to get out of the dialog. We should be rendering these items just as we do the block inserter selections. That works really well and is pretty accessible.

For the category selections, can you add `aria-current="true"` to the active category? This will tell screen readers which one is selected.

If you want an easy way to remember if you need to communicate info to screen readers, ask yourself this question. Do I conditionally add a class to show that the item is selected or active? If so, probably need to communicate through ARIA that way screen reader users have the same info.

Thoughts?

Thanks.

youknowriad commented:
>  We should be rendering these items just as we do the block inserter selections. That works really well and is pretty accessible.

you mean render these after the selection in the DOM right? Unfortunately, that's not possible because of the design : it's a panel that is shown at the right of the inserter. Can we maybe add a live message when we select a category to explain how to navigate and get back (Escape)?

> For the category selections, can you add aria-current="true" to the active category? This will tell screen readers which one is selected.

I can do this

alexstine commented:
@youknowriad An `aria-live` message is the wrong solution here. I mean, does it work? Sure. However, it still will not fix our problem. This is another example of why we should be making the DOM flow. If the DOM flows, there is no need for custom interactions or constant prompts to tell screen readers how to use something. Gutenberg is already noisy enough with these.

When you say design, do you mean visual design or the way the code is currently used right now?

Also, the screen reader does not switch modes on focus. When the screen reader stays in browse mode, users key inputs do not get sent to the application so the list items do not change. Currently, what are you moving focus to? Are you moving it to the list wrapper or the first list item? The wrapping dialog could also be causing this.

1. Escape does work but I think we can do better to allow users more freedom instead of adding yet another custom interaction model based on an announcement users need to listen for.
2. Still need to fix why screen reader is refusing to switch to edit/focus mode.

I'll try to get some more feedback on this from the Accessibility Team as well. It might be good to have others input and suggestions on how to improve this. Maybe something I have not thought of.

youknowriad commented:
> When you say design, do you mean visual design or the way the code is currently used right now?

I mean visual design. Thinking more though, it's the same for dropdown menus, this should probably work just like any regular dropdown menu.

> Currently, what are you moving focus to? Are you moving it to the list wrapper or the first list item? The wrapping dialog could also be causing this.

I think the list wrapper as that might be the default behavior of useDialog

ntsekouras commented on packages/block-editor/src/components/block-draggable/draggable-chip.js:
```diff
@@ -10,7 +10,8 @@ import { dragHandle } from '@wordpress/icons';
 
-export default function BlockDraggableChip( { count, icon } ) {
+export default function BlockDraggableChip( { count, icon, isPattern } ) {
+	const patternLabel = isPattern && __( 'Pattern' );
```
Can we pass the `label` as a param here instead of `isPattern`. We will probably need it for the media tab too.

mtias committed:
Try a different style for active navigation element.

alexstine commented:
@youknowriad Actually, the DOM order really does not seem unreasonable. Looks like that right-side panel does follow pretty closely in DOM order to me. This would be worth a try.
1. Fix the focus to land on the list item. That may make the screen reader behave better.
2. Pass a prop to `useDialog` that can disable `useConstrainedTabbing`.
Or
2. Remove `useDialog` all together because I really do not think it belongs here.
3. The block inserter allows a roving tab order. From memory, it is impossible to tab out of it because it starts the user from the beginning of the inserter again. This is a pattern we should use for this as well but it should be constrained tabbing at the sidebar level, not the list of patterns itself.

Can you commit this in one go? Then it can be reverted if it does not work. I believe that should get us pretty close to our goal.

Essentially, this needs to behave somewhat similar to the list view. It is a list of items after all.

youknowriad committed:
Implement roving tab order for the pattern categories and remove dialog

youknowriad commented:
@alexstine I went ahead and applied your suggestions:

 - Implemented roving tab order for the pattern categories list
 - Removed the dialog and constrained tabbing behavior
 - Fixed the focus when selecting a category

Let me know what you think.

mtias commented:
This is getting closer, nice work. I noticed that pressing escape when the second panel is open doesn't bring focus back to the navigation link that opened the panel. Should it?

youknowriad commented:
> This is getting closer, nice work. I noticed that pressing escape when the second panel is open doesn't bring focus back to the navigation link that opened the panel. Should it?

That was removed with the removal of "useDialog" usage. I can bring only that part if needed but I think @alexstine thought it was not necessary if it's not really a dialog.

alexstine commented:
@youknowriad This is getting really close.

> Implemented roving tab order for the pattern categories list
Works, nice job, 100%!

> Removed the dialog and constrained tabbing behavior
This works, nice job.

> Fixed the focus when selecting a category
Focus is placed correctly now. Nice work.

Further suggestions:
1. When I press Enter to insert a pattern, I hear nothing telling me it was successful. Could you please replicate the announcement as done with inserting a block?
2. The sluggishness is really bad. In Firefox, selecting a category nearly crashes the page. Also, navigating the list of blocks seems really horribly slow. Any idea what is going on here?
3. There are a couple other issues I am currently thinking about in regards to the pattern category navigation. It seems to not be behaving well with screen readers. I'll come back soon with a possible solution to make things more stable. Researching our options.

Thanks.

youknowriad commented:
> When I press Enter to insert a pattern, I hear nothing telling me it was successful. Could you please replicate the announcement as done with inserting a block?

The announcement is already there and I verified that the `speak` call is triggered. So I'm not sure why you're not hearing it. In fact inserting patterns and blocks use the same code. Also I didn't touch that code in this PR so I suspect that you'd have the exact same behavior in trunk.

> The sluggishness is really bad. In Firefox, selecting a category nearly crashes the page. Also, navigating the list of blocks seems really horribly slow. Any idea what is going on here?

That's performance related I guess? What can you tell me more about it that could help, how do you define "sluggishness" here? Regardless, I think it's independent of this PR and shouldn't be a blocker here.

alexstine commented:
@youknowriad Thoughts.

> That's performance related I guess? What can you tell me more about it that could help, how do you define "sluggishness" here? Regardless, I think it's independent of this PR and shouldn't be a blocker here.

I think it should not be a blocker if this PR did not cause it. Anyway, when I open the inserter, things are fine at first. It seems like some type of async rendering makes the page get slower and slower. In Firefox on Windows, the page becomes unresponsive for me like a blocking script of sorts. I really do feel like we should get other Windows users to check this out to make sure we're not introducing a big regression in performance. I cannot check things out further because when Firefox nearly crashes, it makes my screen reader unusable. That is a long outstanding bug that NVDA developers have been working on for some time.

Maybe @tellthemachines can give this one a test?

> The announcement is already there and I verified that the speak call is triggered. So I'm not sure why you're not hearing it. In fact inserting patterns and blocks use the same code. Also I didn't touch that code in this PR so I suspect that you'd have the exact same behavior in trunk.

Do you know if `speak()` is receiving "assertive" as the second argument? I will check this out on trunk.

Thanks.

youknowriad commented:
> Do you know if speak() is receiving "assertive" as the second argument? I will check this out on trunk.

No, it's not. It's this line of code.

https://github.com/WordPress/gutenberg/blob/7ac668cabd4803bd03d40c317b81042864ccfcd7/packages/block-editor/src/components/inserter/hooks/use-insertion-point.js#L130

alexstine commented:
@youknowriad Okay, so the issue with blocks/patterns being inserted is also an issue on trunk. I'll open up a follow-up PR.

The render issue also exists on trunk so I guess that can be looked in to later.

The last issue is the weird one. Here is the related NVDA issue.

https://github.com/nvaccess/nvda/issues/13221

NVDA should switch to edit mode when entering the list, but then it switches to browse mode because the element is a button. Is there any option to keep the current functionality and just use a select/combo box as trunk currently has for the category selections? It is not great UX, but the other solutions are much harder. The other thing we could do is remove the buttons from the list and wrap them in some type of region that way we can attach an `aria-label`.

Thoughts?

youknowriad commented:
> Is there any option to keep the current functionality and just use a select/combo box as trunk currently has for the category selections?

Do you mean, using an actual "select" element? if that's the case, it's probably not possible, the whole purpose of this PR is to apply the new design (list of buttons that open a panel).

> The other thing we could do is remove the buttons from the list and wrap them in some type of region that way we can attach an aria-label.

Can you clarify a bit more here, I'm not sure what you mean, should we use some kind of "role" instead of using buttons? Do you have an example of the desired HTML?

alexstine commented:
@youknowriad Sorry, still trying to work through this. I also have no guidance from spec at this point since nothing seems to work.

One last thing worth trying. After Patterns tab is selected and categories render, can you place focus on the first item just as you do for the patterns within a category? I am not convinced this will work because of the buttons but let's see if we can fool the screen readers for now.

If not, I think we should convert these buttons in to a `<nav>` element with `aria-label="Block Pattern Categories"`. It will be weird, but it will do a couple things.
1. Tell screen reader users they are in a special landmark for category navigation.
2. Add a descriptive ARIA label.
3. Still present as a list.
4. Will allow tabbing in browse mode and once category is selected, should force screen reader to edit mode to navigate the list of patterns in the selected category.

youknowriad committed:
Use nav + list

youknowriad commented:
@alexstine I've prepared two commits:

1- 4034fcd this one uses the "focus first element" approach 
2- 46b3936 this one uses the "nav + list" approach

You can cherry-pick each of them separately and let me know what you think.

alexstine commented:
@youknowriad Looks like you force-pushed the branch? I can't see each commit any longer, just the nav.

```
alexstine@asm gutenberg % git log
commit 46b39363a137255db6f6dfc4a3f19f838348c654 (HEAD -> update/patterns-inserter-design, origin/update/patterns-inserter-design)
Author: Riad Benguella <Redact>
Date:   Wed Oct 12 09:10:07 2022 +0100

    Use nav + list

commit 45a295c2abb29a29aed68226f68c0b0c4f9c495c
Author: Riad Benguella <Redact>
Date:   Fri Oct 7 10:38:25 2022 +0100

    Implement roving tab order for the pattern categories and remove dialog
```

youknowriad commented:
@alexstine yes, it's not on the branch but you can checkout it locally to test: `git checkout 4034fcd3b446b37c20db4f69a10d9d6da645253e`

alexstine commented:
@youknowriad I get this.

```
alexstine@asm gutenberg % git checkout 4034fcd3b446b37c20db4f69a10d9d6da645253e
fatal: reference is not a tree: 4034fcd3b446b37c20db4f69a10d9d6da645253e
```

youknowriad commented:
Can you try `git fetch origin && git checkout 4034fcd3b446b37c20db4f69a10d9d6da645253e` If it doesn't work, I'll prepare a branch for you tomorrow (not on my desk now)

alexstine commented:
@youknowriad Same error. 😞

youknowriad commented:
Ok, so I pushed the commit on a branch, This should work `git fetch && git checkout inserter/focus-first-pattern-category`

alexstine commented:
@youknowriad Okay, awesome.

Focus did not end up working as I thought. You placed focus on the first item of the category list which should have worked, but sadly, it does not work due to how NVDA has changed the screen reader to behave in these situations.

The navigation landmark is a much better idea. Following improvements are needed.
1. Remove the tabindex. Since we are working in buttons, buttons have tabindex 0 by default. Negative tabindex is breaking the ability to tab to the next button since arrow key navigation for screen readers in this context is not possible.
2. Is there a way you can put the Explore all patterns button in the same list?

Other than that, I would be okay doing a final check after this and then green lighting it.

For the future, I guess we're going to have to wrap the inserter in a `role="application"` so that way we can handle the native keyboard events to Gutenberg without one of the screen readers breaking because it is deciding to not follow ARIA spec.

Nice work.

Thanks.

youknowriad commented:
> Remove the tabindex. Since we are working in buttons, buttons have tabindex 0 by default. Negative tabindex is breaking the ability to tab to the next button since arrow key navigation for screen readers in this context is not possible.

I'm not entirely certain what tabindex are you talking about, also it seems that you're saying that we should remove the "arrow key navigation" (Composite items) from here right? Do you think you would be able to make the change as I'm not 100% clear? 

> Is there a way you can put the Explore all patterns button in the same list?

We can, it's close in the component tree, but it does feel very different to me in terms of interaction and behavior, in this sense, it doesn't feel like it belongs to the same list as the "categories".

alexstine commented:
@youknowriad More info.

> I'm not entirely certain what tabindex are you talking about, also it seems that you're saying that we should remove the "arrow key navigation" (Composite items) from here right? Do you think you would be able to make the change as I'm not 100% clear?

You just need to remove the wrapping composite component from the pattern categories.

> It's inspired by the WAI-ARIA Composite Role and implements all the keyboard navigation mechanisms to ensure that there's only one tab stop for the whole Composite element.

This is not necessary since we now have a list of buttons in a `<nav>` tag.

I think moving the Explore all patterns button will help out. I know it is not really a category, but we are trying to communicate relationship between navigation and what will soon be the rendered categories. The WordPress add plugins page follows the same model I think? Showing all plugins but then it is filterable? Even though all is not technically a category, it is still grouped together for better screen reader relationship.

Thanks.

youknowriad committed:
Remove composite items

youknowriad commented:
@alexstine I've removed the composite items, let me know how it feels now. Thanks

alexstine commented:
@youknowriad Okay, time to get picky.

```
<h4 data-wp-c16t="true" data-wp-component="Heading" class="components-truncate components-text components-heading css-1yv8prr e19lxcc00">Featured</h4>
```

Is it really necessary to have a heading here?

The pattern category links need `aria-current="true"`.

After these two issues, we can refine some other stuff in another PR and get this one in. I mainly want to find out why constrained tabbing is acting a little weird with the `role="listbox"` but I would not consider that a blocker since the escape key or tab key can navigate out.

Thanks.

youknowriad commented:
> Is it really necessary to have a heading here?

I'm following the design mockups here, that said, I can definitely make it a regular div or role="presentation" or something like that

youknowriad committed:
Add aria-current

youknowriad commented:
I added the aria-current

alexstine commented:
@youknowriad Regular div instead of the heading would be good. I try to avoid `role="presentation"` as most people do not understand it especially given how limited of a use-case it has. I would vote for clearer code if it is not too much work.

After that, this is good to go.

youknowriad committed:
Remove the heading component

youknowriad commented:
Heading removed.

ntsekouras commented on packages/block-editor/src/components/inserter/style.scss:
```diff
@@ -235,6 +230,104 @@ $block-inserter-tabs-height: 44px;
+		width: 300px;
+	}
+
+	.block-editor-block-card {
```
Nit: I'm following suit for the Media tab here and I think this and the two below styles(`block-editor-block-card__title, h3`) are not relevant here.

alexstine commented:
@youknowriad Signing off on the A11Y review. We'll make further improvements to the UX in the future but this now has know blocks from being tested and eventually released.

Thanks.

ntsekouras reviewed (approved):
🔥 🔥

youknowriad committed:
Remove useless styles

youknowriad committed:
revert unnecessary change

pierre-dickinson commented:
Hi @jameskoster
I'm trying to find a way to reorder the inserter tabs as seen on your screenshot. Please help me find a php or js filter to have this new default tabPanel order : 
1. Patterns
2. Blocks
3. Images

pattern panel from the canvas in first place ?
> 
> <img alt="Screenshot 2022-09-13 at 10 45 09" width="823" src="https://user-images.githubusercontent.com/846565/189869547-a762d49c-e3c2-4f69-bc99-d7ed1592ad3a.png">
